File: go/message_test.go
package confidentiality

import (
	"bytes"
	"crypto/rand"
	"encoding/hex"
	"testing"
)

func TestMessage(t *testing.T) {
	defer func() {
		randomReader = rand.Reader
	}()

	for _, vector := range loadTestVectors(t, "message-vectors.txt", 5) {
		t.Run(string(vector["name"]), func(t *testing.T) {
			var (
				key       = vector["key"]
				iv        = vector["nonce"]
				decrypted = vector["plaintext"]
				encrypted = append(iv, vector["ciphertext"]...)
			)
			t.Run("encrypt", func(t *testing.T) {
				randomReader = bytes.NewBuffer(iv)
				test, err := Encrypt(decrypted, key)
				if err != nil {
					t.Errorf("encrypt failed: %v", err)
				} else if remaining := randomReader.(*bytes.Buffer).Len(); remaining != 0 {
					t.Errorf("%d bytes of random remaining", remaining)
				} else if !bytes.Equal(test, encrypted) {
					t.Errorf("encrypted failed:\ngot:\n%sexpected:\n%s", hex.Dump(test), hex.Dump(encrypted))
				}

			})
			t.Run("decrypt", func(t *testing.T) {
				test, err := Decrypt(encrypted, key)
				if err != nil {
					t.Errorf("decrypt failed: %v", err)
				} else if !bytes.Equal(test, decrypted) {
					t.Errorf("decrypted failed:\ngot:\n%sexpected:\n%s", hex.Dump(test), hex.Dump(decrypted))
				}
			})
		})
	}
}

File: java/src/test/java/io/maze/confidentiality/MessageTest.java
package io.maze.confidentiality;

import static org.junit.Assert.assertThat;
import static org.hamcrest.core.Is.is;
import static org.hamcrest.core.IsEqual.equalTo;

import java.util.Collection;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.Parameterized;
import org.junit.runners.Parameterized.Parameters;
import org.powermock.core.classloader.annotations.PowerMockIgnore;
import org.powermock.modules.junit4.PowerMockRunner;
import org.powermock.modules.junit4.PowerMockRunnerDelegate;

import io.maze.confidentiality.internal.Hex;

@RunWith(PowerMockRunner.class)
@PowerMockRunnerDelegate(Parameterized.class)
@PowerMockIgnore("javax.crypto.*")
public class MessageTest {
    public static final String VECTORS = "../testdata/message-vectors.txt";

    private Helper.Vector vector;

    public MessageTest(Helper.Vector test) {
        vector = test;
    }

    @Parameters(name = "{0}")
    public static Collection<Helper.Vector> vectors() throws Exception {
        return Helper.loadVectors(VECTORS);
    }

    @Test
    public void encryptingAMessage() throws Exception {
        System.out.println("MessageTest.encrypting a message: " + vector);

        final byte[] want = Helper.combine(vector.getBytes("nonce"), vector.getBytes("ciphertext"));
        final byte[] test = Message.encrypt(vector.getBytes("plaintext"), vector.getBytes("key"));

        System.out.println("decrypted:");
        Hex.dump(vector.getBytes("plaintext"));
        System.out.println("encrypted:");
        Hex.dump(test);

        assertThat("encrypt returns byte[]", test instanceof byte[], is(true));
        assertThat("encrypted length is " + want.length, test.length, is(want.length));
    }

    @Test
    public void decryptingAMessage() throws Exception {
        System.out.println("MessageTest.decrypting a message: " + vector);

        final byte[] want = vector.getBytes("plaintext");
        final byte[] test = Message.decrypt(Helper.combine(vector.getBytes("nonce"), vector.getBytes("ciphertext")),
                vector.getBytes("key"));

        System.out.println("encrypted:");
        Hex.dump(Helper.combine(vector.getBytes("nonce"), vector.getBytes("ciphertext")));
        System.out.println("decrypted:");
        Hex.dump(test);      

        assertThat("decrypt retusn byte[]", test instanceof byte[], is(true));
        assertThat("decryption succeeds", test, equalTo(want));
    }
}